rest/.jvm/src/main/scala/io/udash/rest/RestServlet.scala

```diff
@@ -47,27 +47,32 @@ class RestServlet(
 
   override def service(request: HttpServletRequest, response: HttpServletResponse): Unit = {
     val asyncContext = request.startAsync()
+    val completed = new AtomicBoolean(false)
+
+    // need to protect asyncContext from being completed twice because
+    // servlet may reuse the same context instance between subsequent requests (not cool)
+    def completeWith(code: => Unit): Unit =
+      if (!completed.getAndSet(true)) {
+        code
+        asyncContext.complete()
+      }
+
     asyncContext.setTimeout(handleTimeout.toMillis)
     asyncContext.addListener(new AsyncListener {
       def onComplete(event: AsyncEvent): Unit = ()
-      def onTimeout(event: AsyncEvent): Unit = {
-        writeFailure(response, Opt("server operation timed out"))
-        asyncContext.complete()
-      }
+      def onTimeout(event: AsyncEvent): Unit =
+        completeWith(writeFailure(response, Opt("server operation timed out")))
```

Comment:
I think in this case the context is completed by the framework (hence the double completion issue).

Reply:
Double completion was happening because I was completing the request here (due to timeout) and then later when the actual request processing finished on the server.
As far as I understand completing the request here does not hurt (although it may be redundant).

Reply:
My comment comes from the fact that not completing the context here before your changes still triggered the bug. You can even see `onComplete` getting called within this listener.

Reply:
That's right. Current implementation with condition variable protects against this too, right?

Reply:
Yeah, it seems it's safe to keep that call.